homeassistant/components/bmw_connected_drive/__init__.py

```diff
@@ -45,48 +53,155 @@
-    accounts = []
-    for name, account_config in config[DOMAIN].items():
-        accounts.append(setup_account(account_config, hass, name))
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the BMW Connected Drive component from configuration.yaml."""
+    if DOMAIN in config:
+        for entry_config in list(config[DOMAIN].values()):
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN, context={"source": "import"}, data=entry_config
+                )
+            )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up BMW Connected Drive from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
 
-    hass.data[DOMAIN] = accounts
+    # Convert data to dict to remove settings that are now stored as options
+    entry.data = dict(entry.data)
+    default_options = {
+        CONF_READ_ONLY: entry.data.pop(CONF_READ_ONLY, False),
+        CONF_USE_LOCATION: False,
+    }
+
+    # Create options based on user input if no options are stored
+    if list(entry.options) != list(default_options):
+        default_options.update(entry.options)
+        entry.options = default_options
+        hass.config_entries.async_update_entry(entry, options=entry.options)
+
+    try:
+        account = await hass.async_add_executor_job(
+            setup_account, entry, hass, entry.data[CONF_USERNAME]
+        )
+        await hass.async_add_executor_job(account.update)
+    except Exception as ex:
+        raise ConfigEntryNotReady from ex
+
+    hass.data[DOMAIN][entry.entry_id] = account
 
-    def _update_all(call) -> None:
+    async def _async_update_all(service_call=None):
         """Update all BMW accounts."""
-        for cd_account in hass.data[DOMAIN]:
+        await hass.async_add_executor_job(_update_all)
+
+        return True
+
+    def _update_all() -> None:
+        """Update all BMW accounts."""
+        for cd_account in hass.data[DOMAIN].values():
             cd_account.update()
 
+    # Add update listener for config entry changes (options)
+    if weakref.ref(update_listener) not in entry.update_listeners:
+        entry.add_update_listener(update_listener)
+
     # Service to manually trigger updates for all accounts.
-    hass.services.register(DOMAIN, SERVICE_UPDATE_STATE, _update_all)
+    hass.services.async_register(DOMAIN, SERVICE_UPDATE_STATE, _async_update_all)
 
-    _update_all(None)
+    await _async_update_all()
 
-    for component in BMW_COMPONENTS:
-        discovery.load_platform(hass, component, DOMAIN, {}, config)
+    for platform in BMW_PLATFORMS:
+        if platform != NOTIFY_DOMAIN:
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(entry, platform)
+            )
+
+    hass.async_create_task(
+        discovery.async_load_platform(hass, NOTIFY_DOMAIN, DOMAIN, {}, entry.data)
```

Comment:
If I understand it correctly, I don't have the original config that was passed to `async_setup` anymore as this is called from `async_setup_entry`?
Should `notify` be setup differently when using `async_setup_entry`?

Reply:
We can store the original config in hass.data from `async_setup`.

Reply:
Ok. Maybe that's a stupid question, but I though that `async_setup` is only called when loading from configuration.yaml? My quick tests also show the same (nothing of the component's `DOMAIN` available in `async_setup`.

How should the notification platfrom be set up from a config entry, i.e. when adding it via UI?

Reply:
The `async_setup` will be called once per integration regardless of how the setup is started.

Reply:
Use `async_load_platform` to load the notify platform.

Reply:
So, moved `async_load_platform` for notify to `async_setup`. But I unfortunately do not understand how I am able to get the config created through ConfigFlow in here, as this only checks for configuration.yaml content?
`config[DOMAIN]` is throwing a `KeyError` (as there is nothing regarding bmw_connected_drive) in `config`.

And even if it was, the notify platform relies on having a `ConnectedDriveAccount` and the only place where I instantiate an this should be in `__init__.py`. 

If it is required to load NOTIFY from `async_setup`, would it be possible to call `async_setup` from `async_setup_entry` with only the config required for notify? 
Surely this would work somehow but feels like a dirty hack to me...

Reply:
No, use `async_load_platform` from within `async_setup_entry`. Please see other integrations that support a notify platform and config entries.

Reply:
I hope I understood now the way you want me to go - I checked with huawei_lte and now I am sending only the current entry from `hass.data[DOMAIN]` instead of `entry.data`:
```
    await discovery.async_load_platform(
        hass,
        NOTIFY_DOMAIN,
        DOMAIN,
        {},
        hass.data[DOMAIN][entry.entry_id],
    )
```
Or should I send `hass.data[DOMAIN]` completely and loop over the entries in `notify.py`'s `get_service()`?

Reply:
See the tibber integration for how this should be done.

Reply:
Ah many thanks - I haven't looked at that one. But now I also now where my main error was: I thought I somehow need to pass the relevant config entries via discovery, but the platform loads the data itself.